hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclController.java

```diff
@@ -0,0 +1,667 @@
+            PermissionStorage.ACL_TABLE_NAME, getClass().getName(),
+            AccessController.class.getName());
+          throw new TableNotFoundException(
+              "Table " + PermissionStorage.ACL_TABLE_NAME + " is not yet created");
+        }
+      }
+    }
+  }
+
+  @Override
+  public void preStopMaster(ObserverContext<MasterCoprocessorEnvironment> c) {
+    if (checkInitialized()) {
+      hdfsAclHelper.close();
+    }
+  }
+
+  @Override
+  public void postCompletedCreateTableAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      TableDescriptor desc, RegionInfo[] regions) throws IOException {
+    if (!desc.getTableName().isSystemTable() && checkInitialized()) {
+      TableName tableName = desc.getTableName();
+      Path[] paths =
+          new Path[] { pathHelper.getTmpTableDir(tableName), pathHelper.getDataTableDir(tableName),
+              pathHelper.getMobTableDir(tableName), pathHelper.getArchiveTableDir(tableName) };
+      for (Path path : paths) {
+        if (!fs.exists(path)) {
+          fs.mkdirs(path);
+        }
+      }
+      // Add table owner HDFS acls
+      String owner =
+          desc.getOwnerString() == null ? getActiveUser(c).getShortName() : desc.getOwnerString();
+      hdfsAclHelper.addTableAcl(desc.getTableName(), owner);
+      try (Table aclTable =
+          c.getEnvironment().getConnection().getTable(PermissionStorage.ACL_TABLE_NAME)) {
+        SnapshotScannerHDFSAclStorage.addUserTableHdfsAcl(aclTable, owner, desc.getTableName());
+      }
+    }
+  }
+
+  @Override
+  public void postCreateNamespace(ObserverContext<MasterCoprocessorEnvironment> c,
+      NamespaceDescriptor ns) throws IOException {
+    if (checkInitialized()) {
+      Path[] nsDirs = new Path[] { pathHelper.getTmpNsDir(ns.getName()),
+          pathHelper.getArchiveNsDir(ns.getName()), pathHelper.getMobDataNsDir(ns.getName()) };
+      for (Path nsDir : nsDirs) {
+        if (!fs.exists(nsDir)) {
+          fs.mkdirs(nsDir);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void postCompletedSnapshotAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      SnapshotDescription snapshot, TableDescriptor tableDescriptor) throws IOException {
+    if (!tableDescriptor.getTableName().isSystemTable() && checkInitialized()) {
+      hdfsAclHelper.snapshotAcl(snapshot);
+    }
+  }
+
+  @Override
+  public void postCompletedTruncateTableAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      TableName tableName) throws IOException {
+    if (!tableName.isSystemTable() && checkInitialized()) {
+      hdfsAclHelper.resetTableAcl(tableName);
+    }
+  }
+
+  @Override
+  public void postDeleteTable(ObserverContext<MasterCoprocessorEnvironment> ctx,
+      TableName tableName) throws IOException {
+    if (!tableName.isSystemTable() && checkInitialized()) {
+      // remove table user HDFS acl from ns data directory if the user has no global/the ns/other
+      // tables of the ns permissions
```

Comment:
Please describle the logics clearly in your comment here, ex: use a simple example. 
Also we need to remove the mobNsDir's acl ?   the archiveNsDir's acl can not be removed here  because the snapshot reading will depend on it,  please file an sperate issue to address this.